### PR TITLE
Ports: Add patchutils

### PR DESCRIPTION
**Pull request by supercomputer7**
To port this software, I implemented the LibC `strptime` function :)

cc @timschumi

### CODE REVIEW

File: AK/Time.cpp [hunk 1]
```diff
@@ -38,6 +41,299 @@ unsigned day_of_week(int year, unsigned month, int day)
     return (year + year / 4 - year / 100 + year / 400 + seek_table[month - 1] + day) % 7;
 }
 
+#ifndef KERNEL
+Optional<struct tm> convert_formatted_string_to_timespec(StringView input, StringView format, size_t& string_pos)
```
Comment (timschumi) [inline AK/Time.cpp:45]: This looks like a (relatively) pure "move code somewhere else" change, it should probably be a separate commit.

File: AK/Time.h [hunk 2]
```diff
@@ -41,6 +42,12 @@ constexpr bool is_leap_year(int year)
     return year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
 }
 
+#ifndef KERNEL
+Optional<struct tm> convert_formatted_string_to_timespec(StringView input, StringView format, size_t&, size_t&);
```
Comment (timschumi) [inline AK/Time.h:46]: Since the `StringView`s are named already, the so-far unnamed `size_t&` should probably get a name as well.

File: AK/Time.cpp [hunk 3]
```diff
@@ -38,6 +41,299 @@ unsigned day_of_week(int year, unsigned month, int day)
...
+{
+    [[maybe_unused]] size_t format_pos = 0;
```
Comment (timschumi) [inline AK/Time.cpp:47]: Surely, removing the `[[maybe_unused]]` here won't make it complain that `format_pos` is unused, right?

File: AK/Time.cpp [hunk 4]
```diff
@@ -38,6 +41,299 @@ unsigned day_of_week(int year, unsigned month, int day)
...
+    auto tm_or_void = convert_formatted_string_to_timespec(input, format, string_pos, format_pos);
+    if (!tm_or_void.has_value())
+        return {};
+    return tm_or_void.release_value();
```
Comment (timschumi) [inline AK/Time.cpp:51]:
```suggestion
    return tm_or_void;
```

Could also be a return without storing the value in a variable first.

File: AK/Time.cpp [hunk 5]
```diff
@@ -38,6 +41,299 @@ unsigned day_of_week(int year, unsigned month, int day)
...
+Optional<struct tm> convert_formatted_string_to_timespec_restrictively(StringView input, StringView format)
+{
+    size_t string_pos = 0;
+    size_t format_pos = 0;
+    auto tm_or_void = convert_formatted_string_to_timespec(input, format, string_pos, format_pos);
+    if (!tm_or_void.has_value())
+        return {};
+    if (string_pos != input.length() || format_pos != format.length())
+        return {};
+    return tm_or_void.release_value();
```
Comment (timschumi) [inline AK/Time.cpp:63]:
```suggestion
    if (string_pos != input.length() || format_pos != format.length())
        return {};
    return tm_or_void;
```

File: AK/Time.h [hunk 6]
```diff
@@ -41,6 +42,12 @@ constexpr bool is_leap_year(int year)
...
+Optional<struct tm> convert_formatted_string_to_timespec(StringView input, StringView format, size_t&);
+Optional<struct tm> convert_formatted_string_to_timespec_restrictively(StringView input, StringView format);
```
Comment (timschumi) [inline AK/Time.h:48]: If you add a `using` declaration at the bottom of the file (just like the other methods do), then you can drop the `AK::` qualification in other places.

File: Userland/Libraries/LibC/time.cpp [hunk 7]
```diff
@@ -275,6 +276,19 @@ char* asctime_r(const struct tm* tm, char* buffer)
     return buffer;
 }
 
+char* strptime(char const* input, char const* format, struct tm* tm)
+{
+    size_t format_length = strlen(format);
+    size_t input_length = strlen(format);
+    size_t string_pos = 0;
+    auto tm_or_void = AK::convert_formatted_string_to_timespec(StringView(input, input_length), StringView(format, format_length), string_pos);
```
Comment (timschumi) [inline Userland/Libraries/LibC/time.cpp:284]: I'm not sure if there is any actual difference, but we usually do the following instead:

```suggestion
    auto tm_or_void = AK::convert_formatted_string_to_timespec(StringView { input, input_length }, StringView { format, format_length }, string_pos);
```

That said, it might be worth it readability wise to do `StringView format_view { format, strlen(format) };` (`input` respectively) as the very first thing in the function and then just use `.length()` everywhere.

File: Ports/patchutils/package.sh [hunk 8]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env -S bash ../.port_include.sh
+port='patchutils'
+version='0.4.2'
+useconfigure='true'
+files=(
+    "https://github.com/twaugh/patchutils/archive/refs/tags/${version}.tar.gz 2ff95f11946558ce63f4d1167abaccbffd49750152346d5304e03ad884304ad6"
```
Comment (timschumi) [inline Ports/patchutils/package.sh:6]: It seems like there are preconfigured tarballs available at `http://cyberelk.net/tim/data/patchutils/stable/`.

**Review by timschumi — CHANGES_REQUESTED** [review]

**Comment by supercomputer7** [comment]
Let's try this again.

**Comment by timschumi** [comment]
boop, that's on me
